### PR TITLE
#197822 user list pagination

### PR DESCRIPTION
**Pull request by snehalsabade**
PR: https://github.com/tekdi/sunbird-admin-ui/pull/19
Acceptance Criteria

- will Show server side side pagination on table
- show total number of users
- restrict multiple request for single route
- use filter for firstName, lastName, email, phone organization name and status

Steps for Implementation

- create table
- Add feature of pagination from primeng
- Get record length after getting record from API and define total record in table using total Record property
- add filter for firstName, lastName, email, phone organization name and status

Implementation Video :https://www.loom.com/share/a56d15c312aa450e86c6829d849e8d1e?sid=fe8422d5-2001-4e1a-bb36-5c8c30accdf1

### CODE REVIEW

File: src/app/constant.config.ts [hunk 1]
```diff
@@ -8,3 +8,7 @@ export const Roles = [
     { name: 'Public', value: 'PUBLIC' }
 ]
 
+export const Status = [
```
Comment (gouravmore) [inline src/app/constant.config.ts]: This should be part of your component ts file

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts [hunk 2]
```diff
@@ -141,15 +103,34 @@ export class SbUserComponent implements OnInit {
...
+      if (!filters[key]) {
+        delete filters[key]
+      }
+    });
+
+    const body = {
+      request: {
+        filters: filters,
+        limit: event?.rows,
+        offset: event?.first ? (event?.first / 10) + 1 : 0,
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: To show the number of rows - this should be configurable. Add constants for `10`

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts [hunk 3]
```diff
@@ -141,15 +103,34 @@ export class SbUserComponent implements OnInit {
...
     }
 
+    this.userService.loadUserList(body).subscribe(users => {
+      this.organizationsUsersList = users?.result?.response?.content;
+      this.count = users?.result?.response?.count;
+      this.loading = false;
+    }, (error: any) => {
+      this.loading = false;
+      this.messages = [];
+      this.messageService.add({ severity: 'error', detail: error.error.params.errmsg });
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: Use the optional chaining (?.) operator for language constants also

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts [hunk 4]
```diff
@@ -175,4 +156,24 @@ export class SbUserComponent implements OnInit {
...
     })
   }
+
+  onSearch(event: any, column: string): void {
+    this.first = 0
+    if (column === 'organizations' || column === 'status') {
+      this.loadUserList(event);
+    } else if (event.target.value.length > 3) {
+      clearTimeout(this.timeout);
+      let $this = this;
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: You can declare this once - at the start of the method

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 3 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: added

File: src/app/constant.config.ts (hunk 1 above)
Comment (snehalsabade) [reply src/app/constant.config.ts]: resolved

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 4 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: resolved

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 2 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: This is for calculating page number/offset. In this event we not getting offset so we are using it for calculating offset dynamically.

Comment (gouravmore) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: If we want to show - 20 rows in the table - we need to update the code.
I recommend using constants in the codebase that can be easily configured to define the desired number of rows.

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.html [hunk 5]
```diff
@@ -35,14 +31,61 @@ <h5>{{ 'USER_TOTAL' | i18next }} <p-chip label="{{count}}" icon="pi pi-user"></p
...
+                            </span>   
+                        </th>
+                        <th scope="col">
+                            <p-dropdown [options]="status" [filter]="true" [showClear]="true" [(ngModel)]="filteredValue.status"
+                                optionLabel="name" (onChange)="onSearch($event,'status')" placeholder="{{'USER_STATUS_PLACEHOLDER' | i18next}}"
+                                optionValue="value" appendTo="body"></p-dropdown>
+                        </th>
+                        <th scope="col">
+                            <span class="block mt-2 md:mt-0 p-input-icon-left">
+
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.html]: empty span?

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts [hunk 6]
```diff
@@ -141,15 +106,34 @@ export class SbUserComponent implements OnInit {
...
-    editUser(user: OrganizationsUsersList) {
-        this.dialogService.open(AddEditUserComponent, {
-            data: user,
-            header: this.i18nextPipe.transform('USER_EDIT'),
-            width: '30%',
-            height: 'auto'
-        });
+  loadUserList(event: any) {
+    let filters = this.filteredValue;
+    Object.keys(filters).forEach(key => {
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.ts:115]: To enhance the code - we need to utilize the lodash library from the next PR.

File: src/locales/en/en.json [hunk 7]
```diff
@@ -75,7 +76,14 @@
   "USER_BLOCK_SUUCCESSFULLY": "User Block Successfully",
   "USER_UNBLOCK_SUUCCESSFULLY":"User Unblock Successfully",
 
-  "COMMON_USER":"Users",
+  "USER_SEARCH_FIRST_NAME":"Search Firstname",
+  "USER_SEARCH_LAST_NAME":"Search Lastname",
+  "USER_SEARCH_ORGANIZATION":"Search Organization",
```
Comment (gouravmore) [inline src/locales/en/en.json]: Is this language constant used anywhere in PR?

File: src/locales/en/en.json [hunk 8]
```diff
@@ -75,7 +76,14 @@
...
+  "USER_SEARCH_EMAIL":"Search Email",
+  "USER_SEARCH_PHONE":"Search Phone",
+  "USER_SEARCH_STATUS":"Select Status",
```
Comment (gouravmore) [inline src/locales/en/en.json]: Is this language constant used anywhere in PR?

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.html (hunk 5 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.html]: removed

File: src/locales/en/en.json (hunk 8 above)
Comment (snehalsabade) [reply src/locales/en/en.json]: removed

File: src/locales/en/en.json (hunk 7 above)
Comment (snehalsabade) [reply src/locales/en/en.json]: removed unused lang.constant

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 6 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts:115]: Ok sir

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts [hunk 9]
```diff
@@ -55,44 +55,9 @@ export class SbUserComponent implements OnInit {
...
-        this.loading = false;
-      }
-      );
-    });
-  }
-
-  onGlobalFilter(table: Table, event: Event) {
-    table.filterGlobal((event.target as HTMLInputElement).value, 'contains');
+    this.userService.getOrganizations(body).subscribe((response: any) => {
+      this.organizations = response?.result?.response?.content;
```
Comment (gouravmore) [inline src/app/sb-admin/components/pages/sb-user/sbuser.component.ts:60]: error block?

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 2 above)
Comment (gouravmore) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: looks like still it is hardcoded

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 9 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts:60]: Added error block

File: src/app/sb-admin/components/pages/sb-user/sbuser.component.ts (hunk 2 above)
Comment (snehalsabade) [reply src/app/sb-admin/components/pages/sb-user/sbuser.component.ts]: added pageOffsetConstant

**Review by gouravmore — APPROVED** [review]

**Pull request merged**